New Feature: Add an endsWithFilter attribute to includeAll

jasonlyle88 opened the pull request:
## Impact

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes expected existing functionality)
- [X] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Description

Add a new attribute to the incudeAll (ChangeLogIncludeAll class) that can filter an include all path based on the string that ends the name of the file. For example, lets assume a project has the following folder structure:

```
root
    changelog.xml
    packages
        my_pkg_1.pks.sql
        my_pkg_1.pkb.sql
        my_pkg_2.pks.sql
        my_pkg_2.pkb.sql
```

In this instance, we want to control when the package spec and package body are run seperately. While we could prefix everything... that makes it harder for a developer to find when editing. So with a postfix of `.pks.sql` for package specs and a postfix of `.pkb.sql` for package bodies, we can do the following in the root changelot.xml:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<databaseChangeLog
    xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xmlns:ora="http://www.oracle.com/xml/ns/dbchangelog-ext"
    xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-latest.xsd"
>
    <includeAll path="packages" relativeToChangelogFile="true" endsWithFilter=".pks.sql" />
    <includeAll path="packages" relativeToChangelogFile="true" endsWithFilter=".pkb.sql" />
</databaseChangeLog>

```

## Things to be aware of

I have implemented this `endsWithFilter` in the `ResourceAccessor.SearchOptions` class (Thanks @nvoxland for suggesting this class in my previous merge... had no idea I'd be expanding it so soon! 🤣).

I have implemented the endsWithFilter in the default `ResourceAccessor.search` method, as well as `AbstractPathResourceAccessor.search` override method. I didn't think we'd want the filter implemented on any of the other classes that override this default method, but happy to add it to other as deemed appropriate!

Additionally I have implemented this filter to trim whitespace before doing any compare, and to compare case insensitively.

## Things to worry about

Nothing that I can think of.

## Additional Context

The defualt for this filter is the blank string, which does not filter at all.

jasonlyle88 commented:
Hey @filipelautert and anyone else who ends up working on this PR! I have a big ask that is outside the standard way you all work. So PLEASE don't hesitate to say you can't accommodate my request, I will totally understand if that is the case!

I use open source Liquibase that is embedded in another database tool. That database tool is looking to upgrade the open source version of Liquibase in their next release. But they are pulling the version of Liquibase they will be using pretty soon to make sure they get all their testing done. For that reason, I was hoping to be able to get this little bit of functionality into Liquibase 4.23 or 4.24. Otherwise this functionality probably won't end up in the other tool until sometime next year.

I realize this is probably a big ask since its such a fast turn around! But figured I could at least ask just in case. Again, please don't hesitate to tell me if you can't accommodate this quick review and merge process!

filipelautert commented:
Hello @jasonlyle88. If all goes well 4.23.0 will be out this week, so we can work to have this one released with 4.24.0 or any 4.23.x patch - no promises but I believe it is very possible!

jasonlyle88 commented:
Awesome @filipelautert, thank you very much! Early congrats on 4.23, I'm sure that's been quite the effort with the checksum fiasco from 4.22. You guys do a lot of work here, and it is greatly appreciated! 

There is an XSD change since I'm adding a new attribute to includeAll, so I believe that means it can only be included with a major release.

MalloD12 commented on liquibase-standard/src/test/groovy/liquibase/changelog/DatabaseChangeLogTest.groovy:
```diff
@@ -702,4 +712,39 @@ http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbch
+        "2.XML"         | 1
+        "3.XML"         | 1
+        "4.XML"         | 2
+        "5.XML"         | 0
```
The test looks good to me, but it's failing because the negative scenarios are not able to find a file. I think we could create a separate test for it and make sure an exception is thrown, but this is already implemented (see test `includeAll throws an exception when directory not found`).

jasonlyle88 replied:
Sorry, I could have sword I tested this and it had all the test passing, my bad! Regardless,  I would have thought including the `errorIfMissing:false` would take care of that? Would you agree that if the filter provides 0 changesets to execute, no error should be thrown if `errorIfMissing` is set to `false`? If you agree, I'm guessing I need to look into why its throwing that error and not respecting the `errorIfMissing` property

jasonlyle88 replied:
Hmmm, I might need your help with this one @MalloD12. When I debugged that run, I noticed that the DatabaseChangeLog.java file is referencing the variable `errorIfMissingOrEmpty`, which you changed in merge #3901 to be `errorIfMissing`. When this is up to date with master, why would this not be present in this branch?

Appreciate any help here

MalloD12 replied:
Oh, we undo that renaming last Friday for `includeAll` because `errorIfMissingOrEmpty` was still available in the xsd file and was causing confusion. It was a ship blocker a member of the team found on Friday afternoon (see #4433 for more details). So please, use `errorIfMissingOrEmpty`.

jasonlyle88 replied:
AH, that explains everything! Okay, I was worried about that happening, glad to see it got undone just for includeAll. Also explains why I thought I had all the tests working: I did, before that change got pushed through! 🤣

I have updated the test so it uses the appropriate property name and now all the tests are passing

MalloD12 commented:
Hi @jasonlyle88,

Thank you again for extending search options filtering with this new attribute. I really like it!

I just left you two comments, would you mind looking at them? If you agree with them I can either apply the changes or leave them to you.

Daniel.

jasonlyle88 commented on liquibase-standard/src/main/java/liquibase/resource/ResourceAccessor.java:
```diff
@@ -354,5 +365,13 @@ public void setMaxDepth(int maxDepth) {
+
+        public String getEndsWithFilter() { return endsWithFilter; }
+
+        public void setEndsWithFilter(String endsWithFilter) { this.endsWithFilter = endsWithFilter.trim(); }
```
Yeah, I'm good with this

jasonlyle88 replied:
Done in a recent push

jasonlyle88 commented:
@MalloD12 Hey, sorry about the confusion there! Once you told me about the revert from `errorIfMissing` -> `errorIfMissingOrEmpty` for includeAll, it all made more sense.

I renamed the function for setting the `endsWithFilter` to `setTrimmedEndsWithFilter`. Additionally I updated the includeAll groovy test from `errorIfMissing` to  `errorIfMissingOrEmpty` so that the test no longer fails and correctly identifies 0 changesets.

Additionally it looks like you guys recently implemented CodeQL and it called out a Java empty string evaluation I did as insufficient, so I fixed that as well.

Let me know if there is anything else!

MalloD12 commented:
> @MalloD12 Hey, sorry about the confusion there! Once you told me about the revert from `errorIfMissing` -> `errorIfMissingOrEmpty` for includeAll, it all made more sense.
> 
> I renamed the function for setting the `endsWithFilter` to `setTrimmedEndsWithFilter`. Additionally I updated the includeAll groovy test from `errorIfMissing` to `errorIfMissingOrEmpty` so that the test no longer fails and correctly identifies 0 changesets.
> 
> Additionally it looks like you guys recently implemented CodeQL and it called out a Java empty string evaluation I did as insufficient, so I fixed that as well.
> 
> Let me know if there is anything else!

Thanks @jasonlyle88 for getting back to us so quickly with these code changes. My apologies for the confusion of `errorIfMissing` -> `errorIfMissingOrEmpty`, it was my fault because I took that note before we merged that fixed for `includeAll`. I think that's it from my side. I'll approve the PR and move it to the next stage so the rest of the team can look at it too. 

Thanks,
Daniel.

MalloD12 reviewed (approved):
**Approved**.

Again, thank you @jasonlyle88 for another improvement to the existent functionality!

## Review and testing results:

Code changes look good to me. 

Besides a good set of scenarios unit tested provided by @jasonlyle88. I have made a bit of manual testing and it works as expected. See below example:

For the below changelog:
<img width="578" alt="changelog" src="https://github.com/liquibase/liquibase/assets/75833793/64d5ffa9-4c5b-4b6d-b573-6dc47b9fbb1f">

Output was:
<img width="859" alt="update_output" src="https://github.com/liquibase/liquibase/assets/75833793/92643cdf-2a77-40c4-8fb0-32072a3e46c2">

Build and all tests have been executed successfully. Test Harness shows as failed because a job initially failed, but after re-running it passed.
 

### Things to be aware of:
- None

### Things to worry about:
- Not sure if "to worry about" but it would be good to double-check with some manual execution `includeAll` using `minDepth` and `maxDepth` attributes keep working as expected. I relied on existent unit tests for it keep passing, but maybe worth to check that is still ok. cc: @rberezen

jasonlyle88 commented:
Hello all! @filipelautert, @rberezen do you need anything from me for this?

rberezen commented:
@jasonlyle88, I appreciate your valuable contribution to Liquibase. I have a single request for you, which is to display an accurate message when the 'endsWithFilter' fails to match any files. Currently, an error is shown that is linked to the 'errorIfMissingOrEmpty' attribute. The error message states: "Unexpected error running Liquibase: Could not find directory or directory was empty for includeAll 'extra'".

jasonlyle88 commented:
@rberezen Sounds good. I'll get to that sometime in the next couple days

jasonlyle88 commented:
Hey @rberezen, the error message has been updated in my latest commit. Instead of:
`Could not find directory or directory was empty for includeAll '[INCLUDEALL PATH VALUE]'`
The following error message is thrown:
`Could not find directory, directory was empty, or no changelogs matched the provided search criteria for includeAll '[INCLUDEALL PATH VALUE]'`

I chose this more generic path because it could be `minDepth`, `maxDepth`, `endsWithFilter`, or some future addition to includeAll search criteria that cause no changelogs to be found.

MalloD12 commented:
> 

I don't know why (maybe a related merged is related with this) but I see this test failing in GH:

```
Failures: 
Error:    DatabaseChangeLogTest.includeAll throws exception when directory not found:459 Condition not satisfied:

e.getMessage().startsWith("Could not find directory or directory was empty for includeAll '")
| |            |
| |            false
| Could not find directory, directory was empty, or no changelogs matched the provided search criteria for includeAll 'com/example/missing'
liquibase.exception.SetupException: Could not find directory, directory was empty, or no changelogs matched the provided search criteria for includeAll 'com/example/missing'
	at liquibase.changelog.DatabaseChangeLog.includeAll(DatabaseChangeLog.java:753)
	at liquibase.changelog.DatabaseChangeLog.includeAll(DatabaseChangeLog.java:668)
	at liquibase.changelog.DatabaseChangeLogTest.$spock_feature_0_11(DatabaseChangeLogTest.groovy:455)
Caused by: liquibase.exception.SetupException: Could not find directory, directory was empty, or no changelogs matched the provided search criteria for includeAll 'com/example/missing'
	at liquibase.changelog.DatabaseChangeLog.includeAll(DatabaseChangeLog.java:741)
	... 2 more
```

It's weird I run this on Mac and all tests passed locally for me. Have you tried running the tests locally @jasonlyle88? 

Thanks,
Daniel.

jasonlyle88 commented:
I don't think I ran all tests locally when I just changed the error string because it was such a small change 😳. Since it looks like that test is looking for the old error string, I guess my change did in fact break the testing. I will run this all locally and double check this is the case. If it is, I'll update the test with the new expected error string and re-push

MalloD12 commented:
I didn't update my local branch, I did and saw the same test failing for me locally this time. I fixed it and just pushed that change. Thank you, @jasonlyle88 for pointing this out. 

I hope this time it goes well.

jasonlyle88 commented:
Sorry, I'm embarrassed I pulled you in when I should have just seen that!!! thank you for the fix and patience with my haste 😳

MalloD12 commented:
> Sorry, I'm embarrassed I pulled you in when I should have just seen that!!! thank you for the fix and patience with my haste 😳

My pleasure to help you anytime. 

Daniel.

jasonlyle88 commented:
@MalloD12 Sorry to call on you again! Looks like just the "Run Functional Tests" is failing now... but I can't see any details on that. Locally all my tests passed. Not sure if there is something I need to do here or not.

MalloD12 commented:
As far as I know, functional tests have been failing in master for a different reason and the team is already working on it. 

Thanks,
Daniel.

jasonlyle88 commented:
@rberezen @filipelautert with the error message fixed as requested by @rberezen and everything passing to the extent to which I have control, I think I'm done here and ready for the next step!

jasonlyle88 commented:
@MalloD12 @rberezen love the look of that green checkmark! Thanks for your work here!

rberezen reviewed (approved):
As always, @jasonlyle88 , thank you for your dedication and hard work. 
This pull request introduces a new attribute ‘endsWithFilter’ to the ChangeLogIncludeAll class, allowing filtering of includeAll paths based on the ending string of the file name.

- Functional and test harness executions passing.
- New unit test added.
- Manual testing completed.
- Doc ticket created https://datical.atlassian.net/browse/PD-3539

APPROVED

jasonlyle88 commented:
Good morning/afternoon/evening @suryaaki2! I see the merge freeze has been lifted. I'd love to get this in for 4.24 if possible!

jasonlyle88 commented:
@filipelautert @suryaaki2 This has been sitting around for about 2 weeks with no progress, anything I can do here?

Also, it appears the autobuild step of the codeql job is failing - as far as I know this isn't because of anything in my PR. Is this a known issue? Or do you need anythi g for me for this PR?

jasonlyle88 commented:
@filipelautert, thanks for the information. Is there a way to know if you guys are planning a patch release? That way I know not to bug you guys if something isn't getting merged in because of XSD changes!

As for the build failure, I'll definitely take a look and get back to you!

In this case, is there a way for a user such as yourself not to have to remove and add the safe to build tag? Obviously I'll test the build on my local before I say it's fixed... But seems like there is a whole in the process when a merge from develop is pulled in that a build that was previously working can break... But I can't know about it unless I then pull to my local and rebuild!

jasonlyle88 commented:
@filipelautert Sounds good. Just didn't want to be more of a burden than I had to be! I have fixed the issues that arose since the develop merges. If you could remove/add the SafeToBuild label, that would be greatly appreciated!

filipelautert commented:
don't worry, you won't be a burden! Thanks again for all your pr's , and I just switched the label ;)

jasonlyle88 commented:
Thank you! And thanks for being such a great and easy team to work with! Makes it easy to contribute

filipelautert commented:
@jasonlyle88 don't worry, we still have some build issues being fixed. I hope that by today EOD or tomorrow it will be ok then I can rebuild your branch.

filipelautert commented:
@jasonlyle88 fyi the build issues are not resolved yet, will keep you posted.

jasonlyle88 commented:
@filipelautert @sayaliM0412 

Can I ask why this was closed without being merged? I believed this was on track to be merged any time now...

sayaliM0412 commented:
> @filipelautert @sayaliM0412
> 
> Can I ask why this was closed without being merged? I believed this was on track to be merged any time now...

Sorry @jasonlyle88 , I have reopened this again. If we delete the base branch, the PR gets closed automatically. This was due to some testing I am doing on your branch as @filipelautert would have mentioned earlier

filipelautert commented:
Hey @jasonlyle88 - we are using this PR to test the build fixes. Hope you don't mind 👍

jasonlyle88 commented:
@sayaliM0412 No worries, glad it was just a mistake. I saw it opened right as I was hitting my comment button 🤣 sorry I was a little trigger happy!

@filipelautert If I can help you all with your process, happy to be of assistance! Use this PR as needed!

filipelautert commented:
Hello @jasonlyle88 ! I restored some public methods as some extensions may use it. Now I think we don't have any other breaking change.

filipelautert commented:
Thanks @jasonlyle88 ! Build still not fully working but it helped us to find the problems :D Thank you so much for all the help and sorry for taking so long to merge it.
